migrate to new user/group name

kenhys opened the pull request:
for deb, according Debian policy, dynamically allocated name should be prefixed
with an underscore.
(It may be not applicable for group name, but there is an prefixed
group name example e.g. _ssh/_cvsadmin)

See
https://www.debian.org/doc/debian-policy/ch-opersys.html#introduction

Note that adduser doesn't accept _name prefix by default. To avoid
it, we need to specify --force-badname explicitly

  * migration from to td-agent to _fluentd
    * service, hook script, tmpfiles.d and so on.
    * cleanup td-agent statoverride

For rpm, no prefixed underscore for user/group name.

  * migration from td-agent to fluentd
    * service, hook script, tmpfiles.d, permission and so on.
    * cleanup user/group on removing package

kenhys commented:
This PR is focus on deb only.

kenhys commented:
Checking failure.

kenhys commented:
Unexpectedly exception was raised during install  fluentd-apt-source. :thinking: 

```
dpkg: unrecoverable fatal error, aborting:
 unknown system user '_fluentd' in statoverride file; the system user got removed
```

ashie commented:
Still checking...

kenhys commented:
just rebased with recent master.

ashie commented on fluent-package/templates/package-scripts/fluent-package/deb/postinst:
```diff
@@ -7,25 +7,31 @@ set -e
-    if ! getent passwd <%= compat_service_name %> >/dev/null; then
-        adduser --group --system --home /var/lib/<%= package_dir %> <%= compat_service_name %>
+    if ! getent passwd _<%= service_name %> >/dev/null; then
+	if ! getent group <%= compat_service_name %> >/dev/null; then
```
Is it intentional that checking `group` instead of `passwd` here?

kenhys replied:
It should be `passwd`.  thanks.

ashie commented on fluent-package/templates/package-scripts/fluent-package/deb/postinst:
```diff
@@ -7,25 +7,31 @@ set -e
+	    # With underscore prefix, need to disable NAME_REGEX restriction
+	    adduser --group --system --force-badname --home /var/lib/<%= package_dir %> _<%= service_name %>
+	else
+	    groupmod --new-name _<%= service_name %> <%= compat_service_name %>
+	    usermod --login _<%= service_name %> <%= compat_service_name %>
```
Aren't them (user & group) needed to check separately like RPM's one?

kenhys replied:
It should be checked. thanks.

ashie commented on fluent-package/templates/usr/lib/tmpfiles.d/fluentd.conf:
```diff
@@ -1,8 +1,8 @@
-d /tmp/<%= package_dir %> 0755 <%= compat_service_name %> <%= compat_service_name %> - -
+d /tmp/<%= package_dir %> 0755 <%= service_name %> <%= service_name %> - -
```
`_` prefix should be added on Debian & Ubuntu

ashie commented on fluent-package/templates/package-scripts/fluent-package/deb/postinst:
```diff
@@ -7,25 +7,35 @@ set -e
+    dpkg-statoverride --list /etc/<%= compat_package_dir %> >/dev/null || \
+        dpkg-statoverride --update --add _<%= service_name %> _<%= service_name %> 0755 /etc/<%= compat_package_dir %>
+    dpkg-statoverride --list /var/log/<%= compat_package_dir %> >/dev/null || \
+        dpkg-statoverride --update --add _<%= service_name %> _<%= service_name %> 0755 /var/log/<%= compat_package_dir %>
```
I found that the user `td-agent` is remained in /var/lib/dpkg/statoverride even after upgrading to fluent-package.
Is it intentional?

```
td-agent td-agent 755 /var/run/td-agent
td-agent td-agent 755 /var/log/td-agent
td-agent td-agent 755 /etc/td-agent
```

Only the following line is added:
```
_fluentd _fluentd 755 /var/run/fluent
```

kenhys replied:
Reproduced, checking...

kenhys replied:
got it. the conditional expression should be checked correctly...

ashie replied:
Yep, it doesn't check user name.

kenhys replied:
fixed.

kenhys commented:
Upgrade from v4;

statoverride was migrated to fluentd.

```
+ sudo apt install -y ./fluent-package_5.0.0-1_amd64.deb
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'fluent-package' instead of './fluent-package_5.0.0-1_amd64.deb'
The following packages will be REMOVED:
  td-agent
The following NEW packages will be installed:
  fluent-package
0 upgraded, 1 newly installed, 1 to remove and 0 not upgraded.
Need to get 0 B/12.4 MB of archives.
After this operation, 7,632 kB of additional disk space will be used.
Get:1 /home/vagrant/fluent-package_5.0.0-1_amd64.deb fluent-package amd64 5.0.0-1 [12.4 MB]
(Reading database ... 65954 files and directories currently installed.)
Removing td-agent (4.5.0-1) ...
Selecting previously unselected package fluent-package.
(Reading database ... 57271 files and directories currently installed.)
Preparing to unpack .../fluent-package_5.0.0-1_amd64.deb ...
Unpacking fluent-package (5.0.0-1) ...
Setting up fluent-package (5.0.0-1) ...
Installing new version of config file /etc/logrotate.d/td-agent ...
Installing new version of config file /etc/td-agent/td-agent.conf ...
dpkg-statoverride: warning: an override for '/var/run/fluent' already exists, but --force specified so will be ignored
dpkg-statoverride: warning: an override for '/etc/td-agent' already exists, but --force specified so will be ignored
dpkg-statoverride: warning: an override for '/var/log/td-agent' already exists, but --force specified so will be ignored
Failed to preset unit: File /etc/systemd/system/td-agent.service already exists and is a symlink to /dev/null.
/usr/bin/deb-systemd-helper: error: systemctl preset failed on fluentd.service: No such file or directory
Processing triggers for man-db (2.9.4-2) ...
Processing triggers for libc-bin (2.31-13+deb11u5) ...
+ '[' 0 -ne 0 ']'
+ systemctl is-enabled td-agent
masked
+ systemctl is-enabled fluentd
enabled
+ systemctl is-active td-agent
inactive
+ systemctl is-active fluentd
active
+ echo 'CHECK: _fluentd/_fluentd'
CHECK: _fluentd/_fluentd
+ id -u _fluentd
108
+ id -g _fluentd
117
+ sudo dpkg-statoverride --list
root postdrop 2555 /usr/sbin/postqueue
postfix postdrop 2710 /var/spool/postfix/public
td-agent td-agent 755 /var/run/td-agent
root crontab 2755 /usr/bin/crontab
root mlocate 2755 /usr/bin/mlocate
root ssl-cert 710 /etc/ssl/private
_fluentd _fluentd 755 /var/log/td-agent
_fluentd _fluentd 755 /var/run/fluent
root messagebus 4754 /usr/lib/dbus-1.0/dbus-daemon-launch-helper
_fluentd _fluentd 755 /etc/td-agent
root postdrop 2555 /usr/sbin/postdrop
```

kenhys commented:
It requires #492 for Windows CI.

ashie commented on fluent-package/templates/package-scripts/fluent-package/deb/postinst:
```diff
@@ -7,25 +7,43 @@ set -e
+    dpkg-statoverride --list /var/log/<%= compat_package_dir %> >/dev/null || \
+        dpkg-statoverride --update --add _<%= service_name %> _<%= service_name %> 0755 /var/log/<%= compat_package_dir %>
+    dpkg-statoverride --list /var/log/<%= compat_package_dir %> >/dev/null && \
+        dpkg-statoverride --force-all --update --add _<%= service_name %> _<%= service_name %> 0755 /var/log/<%= compat_package_dir %>
```
The following entries is remained after upgrading from td-agent to fluent-package.
Is it intentional?
```
td-agent td-agent 755 /var/run/td-agent
```

kenhys replied:
Oops, it is not intended... I'll fix it.

kenhys commented:
Rebased because  #492 was merged.

kenhys commented:
/var/run/td-agent statoveride configuration was removed.

```
Note, selecting 'fluent-package' instead of './fluent-package_5.0.0-1_amd64.deb'
The following packages will be REMOVED:
  td-agent
The following NEW packages will be installed:
  fluent-package
0 upgraded, 1 newly installed, 1 to remove and 0 not upgraded.
Need to get 0 B/12.4 MB of archives.
After this operation, 7,585 kB of additional disk space will be used.
Get:1 /home/vagrant/fluent-package_5.0.0-1_amd64.deb fluent-package amd64 5.0.0-1 [12.4 MB]
(Reading database ... 65954 files and directories currently installed.)
Removing td-agent (4.5.0-1) ...
Selecting previously unselected package fluent-package.
(Reading database ... 57271 files and directories currently installed.)
Preparing to unpack .../fluent-package_5.0.0-1_amd64.deb ...
Unpacking fluent-package (5.0.0-1) ...
Setting up fluent-package (5.0.0-1) ...
Installing new version of config file /etc/logrotate.d/td-agent ...
Installing new version of config file /etc/td-agent/td-agent.conf ...
dpkg-statoverride: warning: an override for '/var/run/fluent' already exists, but --force specified so will be ignored
dpkg-statoverride: warning: an override for '/etc/td-agent' already exists, but --force specified so will be ignored
dpkg-statoverride: warning: an override for '/var/log/td-agent' already exists, but --force specified so will be ignored
Failed to preset unit: File /etc/systemd/system/td-agent.service already exists and is a symlink to /dev/null.
/usr/bin/deb-systemd-helper: error: systemctl preset failed on fluentd.service: No such file or directory
Processing triggers for man-db (2.9.4-2) ...
Processing triggers for libc-bin (2.31-13+deb11u5) ...
+ '[' 0 -ne 0 ']'
+ systemctl is-enabled td-agent
masked
+ systemctl is-enabled fluentd
enabled
+ systemctl is-active td-agent
inactive
+ systemctl is-active fluentd
active
+ echo 'CHECK: _fluentd/_fluentd'
CHECK: _fluentd/_fluentd
+ id -u _fluentd
108
+ id -g _fluentd
117
+ sudo dpkg-statoverride --list
root postdrop 2555 /usr/sbin/postqueue
postfix postdrop 2710 /var/spool/postfix/public
root crontab 2755 /usr/bin/crontab
root mlocate 2755 /usr/bin/mlocate
root ssl-cert 710 /etc/ssl/private
_fluentd _fluentd 755 /var/log/td-agent
_fluentd _fluentd 755 /var/run/fluent
root messagebus 4754 /usr/lib/dbus-1.0/dbus-daemon-launch-helper
_fluentd _fluentd 755 /etc/td-agent
root postdrop 2555 /usr/sbin/postdrop
```

kenhys commented:
NOTE need to fix CI.

kenhys commented:
Fixed remaining old statoverride for /var/run/td-agent.

ashie commented:
For debian & ubuntu, I confirmed all issues that I found have been fixed. Thanks!
For RPM, I've found some issue:

* When updating from v4, owner of /var/log/td-agent and so on are changed to root:
```
$ sudo LANG=C rpm -Uvh fluent-package-5.0.0-1.el7.x86_64.rpm 
Preparing...                          ################################# [100%]
Migrate group td-agent to fluentd...
failed
Migrate user td-agent to fluentd...
Updating / installing...
   1:fluent-package-5.0.0-1.el7       warning: user td-agent does not exist - using root
warning: group td-agent does not exist - using root
warning: user td-agent does not exist - using root
warning: group td-agent does not exist - using root
################################# [ 50%]
warning: user td-agent does not exist - using root
warning: group td-agent does not exist - using root
warning: user td-agent does not exist - using root
warning: group td-agent does not exist - using root
warning: user td-agent does not exist - using root
warning: group td-agent does not exist - using root
failed
prelink detected. checking /etc/prelink.conf.d/ ...
Removing prelink configuration for td-agent (/etc/prelink.conf.d/td-agent-ruby.conf) ...
Removed symlink /etc/systemd/system/multi-user.target.wants/td-agent.service.
Cleaning up / removing...
   2:td-agent-4.4.2-1.el7             ################################# [100%]
```
* When I remove fluent-pacakge, the user `fluentd` is still remained.
  * It might be originate from td-agent. Removing td-agent also remains `td-agent` user.

kenhys commented:
>  When updating from v4, owner of /var/log/td-agent and so on are changed to root:

I'll check again.

kenhys commented:
Fixed remained user/group issue.
_

kenhys commented:
waiting CI

kenhys commented:
Upgrade from v4:

```
---> Package fluent-package.x86_64 0:5.0.0-1.el7 will be obsoleting
---> Package td-agent.x86_64 0:4.5.0-1.el7 will be obsoleted
--> Finished Dependency Resolution

Dependencies Resolved

==============================================================================================================================================================
 Package                            Arch                       Version                           Repository                                              Size
==============================================================================================================================================================
Installing:
 fluent-package                     x86_64                     5.0.0-1.el7                       /fluent-package-5.0.0-1.el7.x86_64                      64 M
     replacing  td-agent.x86_64 4.5.0-1.el7

Transaction Summary
==============================================================================================================================================================
Install  1 Package

Total size: 64 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
Migrate group td-agent to fluentd...
Migrate user td-agent to fluentd...
  Installing : fluent-package-5.0.0-1.el7.x86_64                                                                                                          1/2 
prelink detected. checking /etc/prelink.conf.d/ ...
Removing prelink configuration for td-agent (/etc/prelink.conf.d/td-agent-ruby.conf) ...
Removed symlink /etc/systemd/system/multi-user.target.wants/td-agent.service.
  Erasing    : td-agent-4.5.0-1.el7.x86_64                                                                                                                2/2 
  Verifying  : fluent-package-5.0.0-1.el7.x86_64                                                                                                          1/2 
  Verifying  : td-agent-4.5.0-1.el7.x86_64                                                                                                                2/2 

Installed:
  fluent-package.x86_64 0:5.0.0-1.el7                                                                                                                         

Replaced:
  td-agent.x86_64 0:4.5.0-1.el7                                                                                                                               

```

Removed:

```
Running transaction
  削除中                  : fluent-package-5.0.0-1.el7.x86_64                                                                                             1/1 
Removing fluentd user...
userdel: fluentd mail spool (/var/spool/mail/fluentd) not found
userdel: fluentd home directory (/var/lib/td-agent) not found
  検証中                  : fluent-package-5.0.0-1.el7.x86_64                                                                                             1/1 

削除しました:
  fluent-package.x86_64 0:5.0.0-1.el7                                                                                                                         

完了しました!
[vagrant@centos7 ~]$ getent passwd fluentd
[vagrant@centos7 ~]$ echo $?
2
[vagrant@centos7 ~]$ getent group fluentd
[vagrant@centos7 ~]$ echo $?
2
```

ashie commented:
I confirmed that RPM related issues are resolved.

Although td-agent user is remained when I remove td-agent manually before upgrading to fluent-package, it's not fluent-package's issue, it's current td-agent's issue (it doesn't remove td-agent user on uninstallation).
